### ISSUE TITLE
RS485 TX-Enable

### ISSUE THREAD

**Issue by lblabr** [jeelabs/esp-link#489]
I try to get your firmware working. i'd like to access my power meter with my smart home. i tried latest release, but i couldn't configure the tx-pin. i never could save the configuration until i used the workaround from here [https://github.com/jeelabs/esp-link/pull/138]

in the ui there are no options for selecting a pin ..... is that just a frontend issue ? may i congfigure the pins another way ?

thanks in advance 

Lars

**Comment by uzi18** [comment]
It is unavailable because rs485 support was removed due to some boot issues.
I can try to build firmware with rs485 for test if you want?

@tve do you have any link to root couse/report of boot issue with rs485? 
Or maybe it is same boot issue we encounter sometimes at first flash on some modules?

**Comment by tve** [comment]
The issue was related to switching to the microsecond resolution timer, but I don't remember the details.

**Comment by uzi18** [comment]
@tve will try to revert it without use of microsecond timer

**Comment by tve** [comment]
I believe it's needed to get the accuracy required by rs485

**Comment by lblabr** [comment]
it seems that there something with my brower cache or browser it self. from another computer and your latest release i could configure tx enable, but i'm not quite sure that it is working ... i could not get some response ... i just ordered an usb stick RS485/USB to monitor the bus traffic, i'm still interested to get some wireless access to the RS485/ModBus...

i disabled debug output to console and boot is fine with wemos mini.... but it would also be great to monitor the bus traffic ether in debug coonsole or debug with uart.... 

do you think its possible to have debug output and bus connected at the same time on different uart's ?

if you like, i will do some debugging, but i'm not a develeoper...

thanks in advance

**Comment by uzi18** [comment]
@lblabr what firmware version did you used?

**Comment by lblabr** [comment]
i started with your the latest release file here:
[https://github.com/jeelabs/esp-link/releases/tag/v3.2.47.alpha](https://github.com/jeelabs/esp-link/releases/tag/v3.2.47.alpha)

**Comment by mj-sakellaropoulos** [comment]
I can confirm the that the boot issue appears only _after_ selecting TX_ENABLE in the UI and rebooting.

According to ESP8266 documentation , the reset is triggered by the hardware watchdog timer.

Maybe microsecond timer settings have interfered in some way with the software WDT ?

If anyone has some insights on this it would be appreciated 🙏🏻 

Here is my uart output
```
▒▒▒U9   1▒gpio0 state=1
TX_ENABLE gpio0 state=1
TX_ENABLE gpio0 state=1
TX_ENABLE gpio0 state=1
ets Jan  8 2013,rst cause:4, boot mode:(3,7)
wdt reset
load 0x40100000, len 2592, room 16
tail 0
chksum 0xf3
load 0x3ffe8000, len 764, room 8
tail 4
chksum 0x92
load 0x3ffe82fc, len 676, room 4
tail 0
chksum 0x22
csum 0x22
2nd boot version : 1.7(5d6f877)
SPI Speed : 40MHz
SPI Mode : QIO
SPI Flash Size & Map: 8Mbit(512KB+512KB)
jump to run user1 @ 1000
▒▒▒g▒s▒▒n|▒$▒l`b▒|{▒l▒n▒▒o▒l`▒▒s▒l▒l
l`▒▒;▒d▒d▒
d`▒▒s▒l
▒▒▒
dl s$▒▒sl▒▒▒c▒▒cp|
cx▒▒▒x▒▒▒c
▒▒'▒no▒
ld▒▒▒▒▒U9       1▒gpio0 state=1
TX_ENABLE gpio0 state=1
TX_ENABLE gpio0 state=1
TX_ENABLE gpio0 state=1
[loops after this]
```

**Comment by uzi18** [comment]
You should avoid gpio0 because it is used on boot, try another gpio please

**Comment by mj-sakellaropoulos** [comment]
OK, I have reworked my board and changed to GPIO2.

Changing PIN assignment in UI for TX_ENABLE causes lockup, and results in the same WDT Reset.

It was also necessary to do complete chip_erase since the parameter was stored in flash and usual flashing procedure does not erase this section.

Using esp-link version 3.2.47 alpha

```
  1471> station:  join, AID = 1
  1471> Wifi AP: station joined, AID = 1
  2270> HTTP GET /pins: 200, 16ms, h=15560
  9730> HTTP GET /menu: 200, 4ms, h=15656
  9739> HTTP GET /pins: 200, 6ms, h=15656
  9750> HTTP GET /wifi/info: 200, 8ms, h=14904
  9751> HTTP GET /system/info: 200, 6ms, h=15664
 15152> Wifi check: mode=AP status=255
 18001> Pins changed: reset=0 isp=-1 conn=-1 ser=-1 swap=0 rx-pup=1 tx_enable=-1
 18001> Serbridge pins: reset=0 isp=-1 tx_enable=-1 swap=0
 18013> SER led=-1
 18013> CONN led=-1
 18953> HTTP POST /pins: 204, 957ms, h=16000
 32000> Pins changed: reset=0 isp=-1 conn=-1 ser=-1 swap=0 rx-pup=1 tx_enable=2
 32001> Serbridge pins: reset=0 isp=-1 tx_enable=2 swap=0 # <==== Here I set TX_ENABLE to GPIO2

#[some time passes here until WDT is triggered]

 ets Jan  8 2013,rst cause:4, boot mode:(3,6)

wdt reset
load 0x40100000, len 2592, room 16 
tail 0
chksum 0xf3
load 0x3ffe8000, len 764, room 8 
tail 4
chksum 0x92
load 0x3ffe82fc, len 676, room 4 
tail 0
chksum 0x22
csum 0x22

2nd boot version : 1.7(5d6f877)
SPI Speed : 40MHz
SPI Mode : QIO
SPI Flash Size & Map: 8Mbit(512KB+512KB)
jump to run user1 @ 1000
```

**Comment by marcosribeirobr** [comment]
@mj-sakellaropoulos the solution to change gpio seems to have not solved, correct?

Is the solution to use the chip with autodirecion control like the MAX13487E or is it possible to think of some software solution?

**Comment by mj-sakellaropoulos** [comment]
> @mj-sakellaropoulos the solution to change gpio seems to have not solved, correct?
> 
> Is the solution to use the chip with autodirecion control like the MAX13487E or is it possible to think of some software solution?

correct, GPIO did not change anything. Solution would be to write software from scratch for wifi-rs485 bridge and bypass esp-link (it is on my TODO list). Chip with autodirection control I assume would work, but I have not tried them.

**Comment by marcosribeirobr** [comment]
The chip with autodirection control I use in other applications. Thanks for the answer.

**Comment by uzi18** [comment]
gpio2 is also bad idea, sorry missed your message

**Comment by mj-sakellaropoulos** [comment]
Since I am using an ESP-01 module, I'm not sure there are any other viable pins?

**Comment by uzi18** [comment]
it is wrong by design, please use other module with more gpios

**Comment by alextrical** [comment]
I'm also looking to get the RS485 implemented, and from testing so far it looks like its the call to the function `tx_enable` in "uart.c"
Line 63 is the cause of the boot loop mentioned in the comment above, not sure why but t does not like 'os_printf' in this 
`os_printf("TX_ENABLE gpio%d state=%d\n", uart0_tx_enable_pin, state);`


![image](https://github.com/jeelabs/esp-link/assets/35117191/eeb3a463-3296-4c7d-995f-542f3e991549)
![image](https://github.com/jeelabs/esp-link/assets/35117191/a11c7211-b9a9-4ddf-b5e1-33416758f79d)


I can reproduce this issue, by commenting in and out Line 63/ `os_printf("TX_ENABLE gpio%d state=%d\n", uart0_tx_enable_pin, state);`

It also still presents the same issue with or without the use of microsecond timer throughout the application. Also I'm compiling with Platformio on the off chance that makes any difference

The only issue I'm having after commenting out the code is that the TX_ENABLE state will change at the beginning of the transmission, but never revert back

**Comment by alextrical** [comment]
> I believe it's needed to get the accuracy required by rs485

I don't believe the microsecond timer is required for RS485, the device most of us are trying to use is the MAX485 . The issue is that the device is Half-Duplex and needs to be told when to Transmit or Receive. 
![image](https://github.com/jeelabs/esp-link/assets/35117191/e94ee0c6-4282-4614-839d-44180d04efcf)
![image](https://github.com/jeelabs/esp-link/assets/35117191/54140040-598a-4eb7-9ce2-9a936f526ebe)



All the IC is really doing is converting TTL 1s and 0s into a differential pair, 
either A high and B low, or A low and B high, it just needs to be told when to switch between sending and receiving 
![image](https://github.com/jeelabs/esp-link/assets/35117191/d706c6a7-e4be-4154-8bc7-8ad89fcd278b)

https://www.cuidevices.com/blog/rs-485-serial-interface-explained
